### ISSUE TITLE
Fix false positives for variables and color functions in `color-function-notation`

### ISSUE THREAD

**Issue by cascornelissen** [stylelint/stylelint#5671]
> Clearly describe the bug

A follow-up from #5638, where the original issue was fixed correctly in my tests but more advanced scenarios still result in false positives, e.g. when using Sass' `color.mix` within `rgba()`.

> Which rule, if any, is the bug related to?

`color-function-notation`

> What code is needed to reproduce the bug?

```scss
@use 'sass:color';

$color: rgba(color.mix(#000, #fff, 35%), 0.6);
```

> What Stylelint configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "color-function-notation": "modern"
  }
}
```

> Which version of Stylelint are you using?

`14.0.1` as well as `main` from commit e2c3473.

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

```
stylelint \"src/**/*.scss\"
```

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, SCSS color functions

> What did you expect to happen?

No errors

> What actually happened (e.g. what warnings or errors did you get)?

```shell
 25:27  ✖  Expected legacy color-function notation  color-function-notation
```

**Comment by randypuro** [comment]
I believe we also get a false positive with an scss color hex code variable and alpha channel combo:

works:
```
color: rgb(#999, 50%);
```

fails:
```
$color-var: #999;
color: rgb($color-var, 50%);
```


hard to know if that's an issue with the scss syntax, or with the rule.

**Comment by jeddy3** [comment]
It's a problem with the rule, not the syntax. The fix can be rolled into this issue as they're closely related.

We can improve the heuristics in the [`isStandardSyntaxColorFunction` util](https://github.com/stylelint/stylelint/blob/main/lib/utils/isStandardSyntaxColorFunction.js). If we change the logic to look for a word starting with `#` __or `$`__ within any of the nodes within the function (including those nested within other functions).

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by lachieh** [comment]
I've added a PR ☝🏻 but there's a question in there

**Comment by lachieh** [comment]
If anyone subbed to this issue is interested, I created a rule proposal about how sass color functions work and how they can potentially be linted to avoid confusion. stylelint-scss/stylelint-scss#582

**Comment by and1-lt** [comment]
It still complains about legacy syntax, if you use namespaced Sass imports :/

```
@use './color';

$black: #000;

$shadowBasic: 0 4px 4px rgb(color.$black, 0.1); // <-- Expected modern color-function notation
$shadowBasic2: 0 4px 4px rgb($black, 0.1); // <-- all good
```

**Comment by jeddy3** [comment]
@and1-lt I can [reproduce using the demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIAkA3AhgJ3gAgwcwCMAKSAGwgwDokDS0wBrAGmwAZKBGASgG4AdAHYgmIAM4BPAQBc0ADwThRo4eAgCAZgEs8C4IOzY+IDAFdSMUUZx6BBg0bRSAtKfNOBEJyYEMPAdwErbClTGCZ9OyNaCEZ3TxgAWwAHKXEgkJMwiPtVcgw4p00BdFJNKCcACxhZdNDw20jVBISYaQLElLSEYLrsw3ATUSkIBKckjAgkmAxUgoTNJSK8J3R89W8wKU01Wsz6u37YMDoMR22BJ2jYjycoEyTSsEcYJzBB4dHxyemti12sho5I4nM5qS7ka6eO4PTRPKQvL5TGaaP7dADafQMGQBBwMNlxjW0HgwMCCGMBBP6kAEohgby2KBe0Mez1ETj8mikFVumnU6mmrWcJUylhAmLsAF1xQBfPoS-aNYGYUEXK4MAqiCoULloATlRE-cROCCMjAYMqonDYhU5dRqZzqNDzUhGm7M2HPdxOy09PZ9Ix2tqO52uzzzRYCZZ4VrTWFOYOaF1OBgwcR+ChQf42-rrASbc6vNCkMAFbyiJL0GDlb6nYYYLP+kC5-Ng0oCGCYJx4U5QFFtG4eGkyPWYcq9kktwLda2NlPidSnFq3OkgrZgm6aZLa3VSBsUowtXtoeMdqQmElepc3by+CABPcHIwCb3lbtlJyYDtsjfFItlB+NG6FhgK0vaRk45Z0rysKcl0Vq9PuIBuvcLLwh+zibkk2ouGYPozohyEwnCLy0uYmwUKK8F+gR8TJLMogQOeIEATkNzVCEx60vMZBqJRvo4oBnhFCU75QIxtAkXQmqvCMLTSHx+GPkhQm-qU+oQKInIFo4hRbjMOHmCx-QbmaMB4GYnZ+BUnIWBWzHTghSkGjMoZeD4-hTlRAk5KRdJ1mMtImGJrxSd+ng3h5RlGL55H5EkgXBTA5hyc417uXenn8dm0VJX5FBOKkUylul97ouKim4vilJGESFCkmVFIEkYbxDCMTi5SlorigYUqNdgsoUvKjZDOa4EDjAfhtvVXnZSA3icsVt6lTNjZFkkFTHsKLweDIa6ZeSTUgEiIHSGg0ZGNmeLdf01QgSkAAKExIr8fEHZSOSTPQsFGNdvUEgNBxDYhOmuC8HSzFNlwwHaJJkuKDikH4aDiKKl3YFVh23TAKRw31SlquYSgfuo8L5KI3qXi8BMWKj122poGBDO4FjwpmYp49gQPvTVeDEtN2BvdzIBoCT0wyc0go-Rzf24gDkqzSDuGU4WtJBEY5B+NMF2rc4oPK+WlbE6TavC4jyNsppkZuFN2vA7rSs3Iyer5eM0OaDUDnUUpituFxsIQOQFztpNRQvCLxvdAjSMo7b+MQuqxwaUslynCB7V0UakMENDdUm+2pqx40aohUn4EEKn20TZD4da5HpvR7TiHF4nlvLOXhvB1nOew3XRYN04CRmFsTg28IjbN+Qrcp4bBtp9nMP81H5sQUs1uh4XOTF98AjJ+3aed6HRu1zgS8owPQ+aCP69j038fGlMO9lxXEF2WHov1r3Ztn63a-thv-RbwfrvZ+s8qbd0XvXc2-9moB3ys2PacRdrnBNgkaA0wpyzR4vkKosgVYQI1rXTBsDsHVBHq0PAXITaam1NApoKV06dCvu2KGC9caHT7lAm+BIMa4iMFjHGDV3r00ZqlFmVYpbvS5tVEAtUe44EFtIoY4hzBtmcJAZouooBdWljKOURCJZtCsjZUBhQaQWhNhwmOXDGgtQ+GMJ6hoGEQ0PvPXOgjeGQKsWjHhSl+G7ncULGu+RbFtWcqkf+h0tAiOZkMcR7NJFoxyLI-mCj2Hv3FilCJHiiiaVgCvK2Lws7xwkZSGWBw5Y9Uwe8Nqh5NDHgrFIUm+14YgAAHpEDRMeAAXhKTpTgumsCcAATglAAKi4EQJwfSBnDIlAAai4KMpAWSfGNBaEoM6ECACisgpibCrNgEJCRsC1LQNgAAjpkDA4hsDPhaMECA2Bs7YBTAQNABA8ElPKXoxsRz7E1lmA0ppbCPHtOmb07pgyRnjMmdMqF8zFnLOsQcVZOR1nk3OnXHZezWaHOqccsJNy7kwAeU8klrz3mfPibLH5iElS1gLG88CJigl5xgAXZFhwVzKgQUy5YJjXFyP6JYxuSl6UqnBDEdUfseJByrofVln9+6D1IMPUes1xW8rvjK2BFwWXvwsV-C2q9CnXw1dyhlYJi46sDi-Q2gqIH50IY2TVjLtUmqYS8BIchlzHB5ecPiHBzV+staqO+HFEzJxtTsJVnDg2rjdVKgoJI7gjjaIHPAG09T-OeiiBSjlFQWolbK5mIdmGKpPp4tkKq1VmpdUWhBJb9UR0rSK-JeBf71Xjf6sETbX4sLcZWp1H9u2hqcZnFx4CQXeyNSs66fDZB3X8fIumwr0mupjYkxoUSmbtliWzX6W6eZ82ndIoJGTJacsOjki07bO2SsYF83EZS7AVM5rNQMDonSJldBTS5EB4R8VPmyKyAoXB0lkqBOJH6NiNtkseA+5aDWxrPjWy+6rGzwILOo71dq04VuFUau9pq-6coDLB7D8G8NgNYXXYdtCsNgm9bg8GE691BFYDBvMCDiVUsrQQkdmGKNggrCSaQVRaTfnlcwm9sBDXKovp6hjwmLiicFBJiw1HTG5MdeysWP8SNdqE9xgs55SBOH-YB+TcbjOTnZNZQD-aCPAdoRUTIy5uwwG2gBlUFiCm0KKLAU6e0ggACZZpzgXN6b8FMgXoNPTkMF3SIX9PhTCqZkLZkLKWXOjmB4aabJNtivyBzIuLhJcS0lzyKUfKeKral3zBqzXMJGLk-TpieDSpyKKIBmPjs9YG2afXIYtYoRUIIHBQuccbKck8jhzwvD7YbZzVbaGzf5PNi8S255Tro3pwTiF1ungW8rOr+C7zOsO1WOpc2zwXl447UCLsSRaA9itK7R5bsnbU+JiwmmTGyd0xyob13jwbbuy8U4kYXg1kcPlZtx9CMNzW6Dr7F4ofRnvtMOHZN+0Ous14mbqPLnTAnUzHDCHpNvxbUj5eaGlNkd68Tq5ZO1FUYRx-VtRGDMM5B59kn1yr7k-Z3j3bQ79u0LYhneMDMmaj3e0pNiAd7PGNfj1m44Yp6gWNOoCCjEMD7yp+rkwCRs75HMGgMCywukdYJ6KwCJuzcDx9a7MACxkHdAACyzQEI7sWNwI1tmtx1-NXtGiEr4-0ATtCI+PednFF77set61Y56gdQrUm09Q4pjDfVUU3UXdjZdAtV0Bll6I-dT7AZHpkbzQdJeOY5HPThy9v1dFNeGrlWK6EQiaAICYNCe8UxSAtv2wHbLged7Iv5Rwvf+9oVh-5DnE-LtKRijPxp5p58w6RDjrT+O9uT8Quv-Ks+t8D5eJZn0LnGcn+CaFMYs-4sBMaEl-pKWZnQomRl1LWXEW5aFnRUKyxV2RKygEOVCmwDvyq3JRgDeVqzQHq3b2rynzymCRGAICKD32XxQ3tx8i738nUUwOfCX1F1oy52R1vwIPylgFEBOigB3HFmIL3wHDBFAR6zv0KH1CfwwGaT6iMDfx6ThVmXS2EJGWyyRW8XnSZw2UxUrWK32XALKCgOoIwBgJeTgMpTOyr36lpTX1UKFzZw0VLWrmQwoJs2PwMNKGF2MIFTF2MglyoOn3yj61Tymj4mm0sOcPyDjzEgTzdjeyylQO73ihgCCk8D9SJm0LrmjycLQICjCOCkiLZB+y5D+1H0NnH0P1X0aE4NCPCPamSkFEj3VguwO30O8ISIKI6mKJLR2l8zrjEhMAkloU4MKkW0QPO01nKNyK1BmCzX1AcRcmTSrG8AYLaC2lD28n6BGmTivyAxACaJaMZ3mjUU6JNliNmi2mTFTHTAwHKGiP4zKNoW2OsKMNw0Q2p0R2A3PlVXQzrUQlOIWHOM4iczMKz2NQKV50bCeJsNwzsPIIcKPyUl+OcFcOl3cI4y2KLHc18OewCP+AiAGmlBAGlCAA). Can you open [a new bug report](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REPORT_A_BUG.yml), please?

**Comment by PhilippeMorier** [comment]
@jeddy3 the issue described by @and1-lt is still [happening](https://stylelint.io/demo/#N4Igxg9gJgpiBcIAkAjAhleACATgcxQApIAbCHAOlRLTAGsAaLABgoEYBKAbgB0A7JHgjRs+ItVqMW7bv36oMAJlEE0xCGUoT6TVp14Ch0ZblWFtUvbL4gGIAM4BPPgBc0ADwQOw9+7fAQfABmAJZ4XsD8WFg8IDgAriQw9rHYkXzR0bFoLgC0CUm5fBC58Xx0xQDufKlYLgkwDFGZsShk9EUlMAC2AA4ujrX18Y3NWQGanbkhfABuaCQhULkAFjDuQw1NGS0B3d0wrlM9-YMIdVtjMeDx9i4Q3bm9OBC9MDgDU90hvjN4ufMcLkgmUwC4QoFNiNtplrrAwDQcDkIXxcm0IB1irkoPFeoswDkYLkwLd7o9nq93uDklDRjtxvDEcjAmj2nQpji8SECS4iRS3h8QjTzgBtK7RYZ02HRdLS3ZhYo4GC1MX0uXXSB8ewwEng2ZEzn4wn2XKVEIuFbYkJBILvQ55eYkEYpEDizIAXTdAF8ru6YbtGWgkeCWejMSV7CtyBa0HxlvyqY5chB9TgcEthdhJf7xkFAnkgmhviQk1jDdzCUUi5mLtCrrE80dC8XSyVvr8+P88Id3tzgUWQiXcnQYI5KuQoLSc9cQXwwSjiQswFMyvZerQYMtKcHyFP6yBZ-OWYs+DAg7k8EioEKjljilq3HGg8tr0qjzVztn9yPHEEkQdsR1JkQ1RLEQj6aNYxcPc1ViA5rzQYEzxceIlSrACsTKCoIGqGDYViPhq2WS8llyIMzxNMC5gWJY8N2MtkjAQ5r07XI1x1a1uXNM4s0uWCQDLXEjV5Mi8nA3po3yRIay-fjBK5HkiW1JIwXIF1eLrOSuj6T57AgVCmLo8YsXWepEO1b5SECdTayleiShmR1SKgfS2iUmhI2JB4DlcGzZPwgSHOoxZ4wgexzQXHJpggj4pKSIzrjAtMYDwRJz0qFZzWSddDM-PiAoTD5W1Kcoqg-DS7PGZSdXuIFem1eIXOJDzKJKLCyoS2JqtUuqGqamAkh8vJMNKnDyts6cuoGmryFyAY3hXUbcNFN1-OlWV1ViBVyGVFa1TlWISTuB5cmmoaXTdaJPX2rAfTVP19zudNWLvGBKhPXaKsmkAynNRbsOWr79wWXoVkQx0Rk6NwQJVN1YgFJjXDQbtYmnGVLuudYmP6AAFF4BWpGzVXVXZXlobjYgx665Tu2EHv4qKCiJE5Pg+tEYDzJVYZu7ISEqNBHBdNGsA2g6QCxmB+m5knxjDOgkl8Migl5IF7GrdCiTlhWhYx3MQhwO4imSXlJ1dG6ruF8Ztq5vaZeuNBlfeLz9ntSnzawanpVpj1vsZ6SNcXbValiMhKneVHgbyJmA7XDclZV4OQAWfnBbYv5Cg+iOGaj-2sX1ONZueDmQg2PLNICv3Cgs7kNBZU93pmIkHYT85eZTnX+Ll5qwr+NEkSY06dKTNmUA5nbE9PVMs4CruER71iUH7ol67Z5vw9bpO+YFjuZ7ZbvwoXpejYb092c5z77a31PukScFckz2x91nsgD-+Re49jgfR-PxPk+3tPOwZ0btPXYXdKR8F7u-AeK9G7x3XtgNu-8b4kDvg-b6YC3gQMPh-HKTdHY4F-lfE0r8gGnhAbLPe4DIFH0-prMeNsEGb3buQjUtcgSHhAlDZk404LQHeB+b6VkgRrHcIHC+IccLr0EWw1Y6x76HDwBaROkZowsMOt5e0g9Tj31gd-cettpSIMFiw9G7tYgSylgYkmDZ9aG1PHcTcbsSb02sSAa2F9iYyy6gMJIJ48iQH2LGKAF13ae1hN7C2+4AlDVNJlXktDphagzIQ5hj9+JHTJE8fGiYtGs10fQjxcMmHbxMSLDG5j3DY2glYzaSd8HElJCdQqAxSkBVCAbYaxtHFm2cZba47jpZeLqSrZ2Q1WnyiSbAABeBSGazZE49UYTMgRI9oIxpjx4IhEQuuFwKtxqeICgAPUICKRCAAvd0pzchnOYLkAAnO6AAVBwQguQrk3Pue6AA1BwR5SBSmi0MSAA4vhkbiJAAAUXcG8MEm4sAZIeFgTZaAsAAEcRg4EcFgQiBw6gQCwKPLAI50AoDEQs8Jvo1nHXJNkoqTwch7MGUck55zLnnNuQ855rz3kcu+b8-5aS5SAoCiCtWKMN5QphSbeF6ysDNKxTimAeKCVKuJWgUlBIg49K9pS-cgZgwLnQKxBJa8CEb0nlIvVQEgzcLRLGf4CS9EMMvqk76+rbVd2rlZUCb1V74JSUg2+IQdFkMFQGa1BrQx7y9bXVEJr-Ubz-qnEhRI0FWoRDazhnqeg1zIHG3BZ99GMItWat1EaPXRvTqm2B3QPCAQzZGrUtQ2BlobRWjE7IzKDl7jGvNAbjFhoZOWrNe8sRKhxE+I4ea8BgzjFk7chNOogHdZw71x8-Ut0YUmk0yDUHAMHXCYdC413xs3S6-+KaQ27VbcBY9sa2IFqdeCktajl1HpZCzYe+Sf41PGNugF5TxaVMltU7ABzammvrbeyEB6xbtLsV002VM+lbTwIqQp7s-31Oia7WDQKZjhSmZeke8ztVyiWdEFZLjdiNgLAOIcircjoogPE-tJoMp2nyDqDRcZunfQ4XewJ67YGmrY7kXdwa038QEyyaJ5kC2icTUQ6ZszX0ydRHJh9ccn0TxgFPA9DZQScNraIz9V6bLMH40ZhcjHNXgtDpa6T1mWTriVK4NY2pKK+tgQR5JSn27iaDVetTznUSuftB55IWmB6+dgLp1MKnq2hqs3OThqESBMfiCxmsRid40dC7ErKp74HnoHd9FYkNYCXhgMvFj3Df6AOvfuGYsAkYw3OIob6P4-zVkourHZjLf3XGOe8tl1zeVcreeyz5Py-kAbMcC5IYrwWSpqnC7r-4lWKuVYStVGq0Bau9Lq-iSROwWmue8EoI1zRLpM7kr99jm3fTu2zU7iiVjNsUJZ-cyKkI5FQkSE9CmE1bqIa+37tp-toSB9pgp8XHMis3Fsv7KE0K2YO-ZyRpaftI8QpD1Hy91Z52YoXJUoRS5A34hD5CAP6VuYtMkKLCTYvPr0wj3Y1OodEiRJ2Ik24cizWK2a0HrqccIRR7Tnn3ZkwCgF6rR9cP-MlIM4t8X6L3hfsNppmBp9FMi8DSgyT+7nu48yxrnRWvvLybjnr0rxCq3BZV799XmKLf+Kt9FuhP7i1s+x1pe7wJbF5AfpTgKJkNCFfibgpdWJ2yv1OnOiAQQ2L6RwNA7zobQ-0XiN0UeQIkgYF7mcy7bHX18Bz3n8Tdai5gB+CiWoAAWb65fc9OyxF2k8-xi8vD8vlXY8qyUbwc37gqtLPjE4LnVMnJcl3RzMyPRXYGinbsC4bx3wthW7AsaBrA4G4NB6Ng4pDoSUNuLQ0W3fut7bYZ49UqmR37rfW6rVUS9QQgoHiCJKBI4XDEILSz+HEfXYZ-WaBldMD-ESfnF-IXQA19EAoEMA9-T-PnWXaAhXb3RKX3OA6aHqV-cA5AzLbLGyXLbAlSF-DNRWAbfhJlXYEbVlHlT5SbBgh5WbAVDfQDUVMFROVbWFKAeFFqLAeAnbVVGAElQfB-OmJ-HA8gh4FAGYOXT3OBYXO3UgmaIEAJOQwiNA2HDAkglXeAwCewRGKAKCZ2TQhQu8FkWhJdAw0iKgnAfZIpOg65MbD5TlF5KbcbGbflebVxTg8VRhHg6VJYQQ6Q8gYQolUQ9VcQm6KjKQsg2aRYS3ITHXPBM9PQ+ItQt3UZa3L+RfTA-TTI3Au7efRuCzIol-CfFyKfYuCnCaR6MI3qGARqEoCgk0OzROYfVQ3A+qZopqNounCLRnP-OOAA81LA-Qxop4PqLoQaTRDoofLHbol-Xolo06OYo4NdYoaGevDeFyeINyZY2aeaQHDHTopY-QqMD4WdeMMfYqcdMoEwo4CGGSPvKqN-ViZjVjPY1yeKFXX6fxM4xYsOIA8YF44cUcccHAZYBYxhLolXcEpI93FIjPNIkrXLVfPdZLfcREn4ZE2tRQ23DE4jY3HEhYSGJEnIxQnTcYwoskp0atQ2EooeczWob7ficEqo0nWo2kZoO6L0EAL0IAA), correct?

**Comment by jeddy3** [comment]
Looks to be. Can you open [a new bug report](https://github.com/stylelint/stylelint/issues/new?assignees=&labels=&template=REPORT_A_BUG.yml) specific to namespaced imports, please?

**Comment by PhilippeMorier** [comment]
:heavy_check_mark: #6610 :eyes:

**Comment by boris-petrov** [comment]
@jeddy3 I just noticed that `#bad3: rgba(var(--bs-primary-rgb), 0.1);` also leads to a warning - and that's also a false-positive. Is that the same as above or should I open a new issue for it?
